File: server/meters/catalog.go
// Package meters owns the run-scoped Company pressure-meter catalog and exact integer rules.
// It deliberately does not import the economy ledger: meters are consequences, never payment.
package meters

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"regexp"
	"sort"

	"cloud-clicker/server/multiplier"
)

const (
	CatalogSchemaVersion = 1
	MinimumValue         = 0
	MaximumValue         = 100
)

var (
	ErrInvalidCatalog = errors.New("invalid meter catalog")
	idPattern         = regexp.MustCompile(`^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$`)
	requiredMeterIDs  = []string{
		"doom.probability",
		"trust.employees.grievance", "trust.employees.standing",
		"trust.investors.grievance", "trust.investors.standing",
		"trust.press.grievance", "trust.press.standing",
		"trust.regulators.grievance", "trust.regulators.standing",
		"trust.users.grievance", "trust.users.standing",
	}
)

type InputKind string

const (
	InputLedgerFact       InputKind = "ledger_fact"
	InputContributionSlot InputKind = "contribution_slot"
)

type Band struct {
	ID         string
	FloorValue int
}

type Input struct {
	Kind                 InputKind
	FactKind             string
	Slot                 multiplier.Slot
	SourceID             string
	Delta                int
	DeltaPerAttendedHour int
}

type Decay struct {
	TowardValue int
	RatePerHour int
}

type Meter struct {
	ID           string
	InitialValue int
	Bands        []Band
	Inputs       []Input
	Decay        *Decay
}

type TrustReseed struct {
	BaseValue            int
	NotorietyNumerator   int64
	NotorietyDenominator int64
	FloorValue           int
	CeilingValue         int
}

type Catalog struct {
	TrustReseed TrustReseed
	Meters      []Meter
	byID        map[string]Meter
}

type rawCatalog struct {
	SchemaVersion int            `json:"schema_version"`
	TrustReseed   rawTrustReseed `json:"trust_reseed"`
	Meters        []rawMeter     `json:"meters"`
}

type rawTrustReseed struct {
	BaseValue            int   `json:"base_value"`
	NotorietyNumerator   int64 `json:"notoriety_numerator"`
	NotorietyDenominator int64 `json:"notoriety_denominator"`
	FloorValue           int   `json:"floor_value"`
	CeilingValue         int   `json:"ceiling_value"`
}

type rawMeter struct {
	ID           string     `json:"id"`
	Scope        string     `json:"scope"`
	MinValue     int        `json:"min_value"`
	MaxValue     int        `json:"max_value"`
	InitialValue int        `json:"initial_value"`
	Bands        []rawBand  `json:"bands"`
	Inputs       []rawInput `json:"inputs"`
	Decay        *rawDecay  `json:"decay"`
}

type rawBand struct {
	ID         string `json:"id"`
	FloorValue int    `json:"floor_value"`
}

type rawInput struct {
	Kind                 InputKind `json:"kind"`
	FactKind             string    `json:"fact_kind,omitempty"`
	Slot                 string    `json:"slot,omitempty"`
	SourceID             string    `json:"source_id,omitempty"`
	Delta                *int      `json:"delta,omitempty"`
	DeltaPerAttendedHour *int      `json:"delta_per_attended_hour,omitempty"`
}

type rawDecay struct {
	TowardValue int `json:"toward_value"`
	RatePerHour int `json:"rate_per_attended_hour"`
}

func LoadCatalog(data []byte) (*Catalog, error) {
	var raw rawCatalog
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&raw); err != nil {
		return nil, fmt.Errorf("%w: decode: %v", ErrInvalidCatalog, err)
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		return nil, fmt.Errorf("%w: trailing JSON", ErrInvalidCatalog)
	}
	if raw.SchemaVersion != CatalogSchemaVersion || len(raw.Meters) != len(requiredMeterIDs) || !validReseed(raw.TrustReseed) {
		return nil, ErrInvalidCatalog
	}
	catalog := &Catalog{
		TrustReseed: TrustReseed{
			BaseValue: raw.TrustReseed.BaseValue, NotorietyNumerator: raw.TrustReseed.NotorietyNumerator,
			NotorietyDenominator: raw.TrustReseed.NotorietyDenominator, FloorValue: raw.TrustReseed.FloorValue,
			CeilingValue: raw.TrustReseed.CeilingValue,
		},
		Meters: make([]Meter, 0, len(raw.Meters)), byID: map[string]Meter{},
	}
	for index, source := range raw.Meters {
		if source.ID != requiredMeterIDs[index] || source.Scope != "company" || source.MinValue != MinimumValue ||
			source.MaxValue != MaximumValue || source.InitialValue < MinimumValue || source.InitialValue > MaximumValue ||
			len(source.Bands) == 0 || source.Inputs == nil || catalog.byID[source.ID].ID != "" {
			return nil, fmt.Errorf("%w: meters[%d]", ErrInvalidCatalog, index)
		}
		meter := Meter{ID: source.ID, InitialValue: source.InitialValue, Bands: make([]Band, 0, len(source.Bands)), Inputs: make([]Input, 0, len(source.Inputs))}
		previousFloor := -1
		seenBands := map[string]bool{}
		for bandIndex, band := range source.Bands {
			if !idPattern.MatchString(band.ID) || seenBands[band.ID] || band.FloorValue < MinimumValue || band.FloorValue > MaximumValue ||
				band.FloorValue <= previousFloor || bandIndex == 0 && band.FloorValue != MinimumValue {
				return nil, fmt.Errorf("%w: meters[%d].bands[%d]", ErrInvalidCatalog, index, bandIndex)
			}
			seenBands[band.ID], previousFloor = true, band.FloorValue
			meter.Bands = append(meter.Bands, Band{ID: band.ID, FloorValue: band.FloorValue})
		}
		seenInputs := map[string]bool{}
		for inputIndex, sourceInput := range source.Inputs {
			input, uniqueness, err := parseInput(sourceInput)
			if err != nil || seenInputs[uniqueness] {
				return nil, fmt.Errorf("%w: meters[%d].inputs[%d]", ErrInvalidCatalog, index, inputIndex)
			}
			seenInputs[uniqueness] = true
			meter.Inputs = append(meter.Inputs, input)
		}
		if source.Decay != nil {
			if source.Decay.TowardValue < MinimumValue || source.Decay.TowardValue > MaximumValue || source.Decay.RatePerHour < 1 || source.Decay.RatePerHour > MaximumValue {
				return nil, fmt.Errorf("%w: meters[%d].decay", ErrInvalidCatalog, index)
			}
			meter.Decay = &Decay{TowardValue: source.Decay.TowardValue, RatePerHour: source.Decay.RatePerHour}
		}
		catalog.byID[meter.ID] = meter
		catalog.Meters = append(catalog.Meters, meter)
	}
	return catalog, nil
}

func validReseed(value rawTrustReseed) bool {
	return value.BaseValue >= MinimumValue && value.BaseValue <= MaximumValue && value.NotorietyNumerator >= 0 &&
		value.NotorietyDenominator > 0 && value.FloorValue >= MinimumValue && value.FloorValue <= value.CeilingValue &&
		value.CeilingValue <= MaximumValue && value.BaseValue >= value.FloorValue && value.BaseValue <= value.CeilingValue
}

func parseInput(source rawInput) (Input, string, error) {
	switch source.Kind {
	case InputLedgerFact:
		if !idPattern.MatchString(source.FactKind) || source.Slot != "" || source.SourceID != "" || source.Delta == nil ||
			source.DeltaPerAttendedHour != nil || *source.Delta == 0 || *source.Delta < -MaximumValue || *source.Delta > MaximumValue {
			return Input{}, "", ErrInvalidCatalog
		}
		return Input{Kind: source.Kind, FactKind: source.FactKind, Delta: *source.Delta}, "fact\x00" + source.FactKind, nil
	case InputContributionSlot:
		if !multiplier.ValidSlot(multiplier.Slot(source.Slot)) || !idPattern.MatchString(source.SourceID) || source.FactKind != "" ||
			source.Delta != nil || source.DeltaPerAttendedHour == nil || *source.DeltaPerAttendedHour == 0 ||
			*source.DeltaPerAttendedHour < -MaximumValue || *source.DeltaPerAttendedHour > MaximumValue {
			return Input{}, "", ErrInvalidCatalog
		}
		return Input{Kind: source.Kind, Slot: multiplier.Slot(source.Slot), SourceID: source.SourceID, DeltaPerAttendedHour: *source.DeltaPerAttendedHour},
			"slot\x00" + source.Slot + "\x00" + source.SourceID, nil
	default:
		return Input{}, "", ErrInvalidCatalog
	}
}

func (catalog *Catalog) Meter(id string) (Meter, bool) {
	if catalog == nil {
		return Meter{}, false
	}
	value, ok := catalog.byID[id]
	return cloneMeter(value), ok
}

func (catalog *Catalog) MeterIDs() []string {
	if catalog == nil {
		return nil
	}
	ids := make([]string, 0, len(catalog.Meters))
	for _, meter := range catalog.Meters {
		ids = append(ids, meter.ID)
	}
	return ids
}

func (catalog *Catalog) ValidateResourceSeparation(resourceIDs []string) error {
	if catalog == nil {
		return ErrInvalidCatalog
	}
	resources := append([]string(nil), resourceIDs...)
	sort.Strings(resources)
	meters := catalog.MeterIDs()
	left, right := 0, 0
	for left < len(resources) && right < len(meters) {
		if resources[left] == meters[right] {
			return fmt.Errorf("%w: meter/economy ID collision %q", ErrInvalidCatalog, resources[left])
		}
		if resources[left] < meters[right] {
			left++
		} else {
			right++
		}
	}
	return nil
}

func cloneMeter(source Meter) Meter {
	result := source
	result.Bands = append([]Band(nil), source.Bands...)
	result.Inputs = append([]Input(nil), source.Inputs...)
	if source.Decay != nil {
		decay := *source.Decay
		result.Decay = &decay
	}
	return result
}

File: client/src/meters/catalog.ts
export const METER_CATALOG_SCHEMA_VERSION = 1 as const;
export const METER_MIN_VALUE = 0 as const;
export const METER_MAX_VALUE = 100 as const;

export type MeterInput =
  | Readonly<{ kind: "ledger_fact"; factKind: string; delta: number }>
  | Readonly<{ kind: "contribution_slot"; slot: MeterSlot; sourceId: string; deltaPerAttendedHour: number }>;

export type MeterSlot = "upgrades" | "milestones" | "faction" | "doctrine" | "commons" | "trust" | "event_buffs" | "prestige";

export interface MeterBand { readonly id: string; readonly floorValue: number }
export interface MeterDecay { readonly towardValue: number; readonly ratePerHour: number }
export interface MeterDefinition {
  readonly id: string;
  readonly initialValue: number;
  readonly bands: readonly MeterBand[];
  readonly inputs: readonly MeterInput[];
  readonly decay: MeterDecay | null;
}
export interface TrustReseed {
  readonly baseValue: number;
  readonly notorietyNumerator: number;
  readonly notorietyDenominator: number;
  readonly floorValue: number;
  readonly ceilingValue: number;
}
export interface MeterCatalog {
  readonly schemaVersion: 1;
  readonly trustReseed: TrustReseed;
  readonly meters: readonly MeterDefinition[];
  readonly byId: ReadonlyMap<string, MeterDefinition>;
}

const mechanicalId = /^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$/;
const slots = new Set<MeterSlot>(["upgrades", "milestones", "faction", "doctrine", "commons", "trust", "event_buffs", "prestige"]);
export const REQUIRED_METER_IDS = Object.freeze([
  "doom.probability",
  "trust.employees.grievance", "trust.employees.standing",
  "trust.investors.grievance", "trust.investors.standing",
  "trust.press.grievance", "trust.press.standing",
  "trust.regulators.grievance", "trust.regulators.standing",
  "trust.users.grievance", "trust.users.standing",
] as const);

function syntax(message: string): never { throw new SyntaxError(message); }

function exactObject(value: unknown, keys: readonly string[], label: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) syntax(`${label} must be an object`);
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index])) syntax(`${label} has invalid keys`);
  return value as Record<string, unknown>;
}

function integer(value: unknown, minimum: number, maximum: number, label: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < minimum || (value as number) > maximum) syntax(`${label} is invalid`);
  return value as number;
}

function identifier(value: unknown, label: string): string {
  if (typeof value !== "string" || !mechanicalId.test(value)) syntax(`${label} is invalid`);
  return value;
}

function parseInput(value: unknown, label: string): { readonly input: MeterInput; readonly uniqueness: string } {
  const discriminator = value !== null && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>).kind : undefined;
  if (discriminator === "ledger_fact") {
    const row = exactObject(value, ["kind", "fact_kind", "delta"], label);
    const delta = integer(row.delta, -METER_MAX_VALUE, METER_MAX_VALUE, `${label}.delta`);
    if (delta === 0) syntax(`${label}.delta must be nonzero`);
    const factKind = identifier(row.fact_kind, `${label}.fact_kind`);
    return { input: Object.freeze({ kind: "ledger_fact", factKind, delta }), uniqueness: `fact\0${factKind}` };
  }
  if (discriminator === "contribution_slot") {
    const row = exactObject(value, ["kind", "slot", "source_id", "delta_per_attended_hour"], label);
    if (typeof row.slot !== "string" || !slots.has(row.slot as MeterSlot)) syntax(`${label}.slot is invalid`);
    const sourceId = identifier(row.source_id, `${label}.source_id`);
    const deltaPerAttendedHour = integer(row.delta_per_attended_hour, -METER_MAX_VALUE, METER_MAX_VALUE, `${label}.delta_per_attended_hour`);
    if (deltaPerAttendedHour === 0) syntax(`${label}.delta_per_attended_hour must be nonzero`);
    return { input: Object.freeze({ kind: "contribution_slot", slot: row.slot as MeterSlot, sourceId, deltaPerAttendedHour }), uniqueness: `slot\0${row.slot}\0${sourceId}` };
  }
  return syntax(`${label}.kind is invalid`);
}

export function loadMeterCatalog(bytes: string | Uint8Array): MeterCatalog {
  const source = typeof bytes === "string" ? bytes : new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  let value: unknown;
  try { value = JSON.parse(source); } catch (error) { syntax(`meter catalog is invalid JSON: ${String(error)}`); }
  const root = exactObject(value, ["schema_version", "trust_reseed", "meters"], "meter catalog");
  if (root.schema_version !== METER_CATALOG_SCHEMA_VERSION || !Array.isArray(root.meters) || root.meters.length !== REQUIRED_METER_IDS.length) syntax("meter catalog has invalid version or meter count");
  const reseedRow = exactObject(root.trust_reseed, ["base_value", "notoriety_numerator", "notoriety_denominator", "floor_value", "ceiling_value"], "meter catalog.trust_reseed");
  const trustReseed = Object.freeze({
    baseValue: integer(reseedRow.base_value, 0, 100, "trust_reseed.base_value"),
    notorietyNumerator: integer(reseedRow.notoriety_numerator, 0, Number.MAX_SAFE_INTEGER, "trust_reseed.notoriety_numerator"),
    notorietyDenominator: integer(reseedRow.notoriety_denominator, 1, Number.MAX_SAFE_INTEGER, "trust_reseed.notoriety_denominator"),
    floorValue: integer(reseedRow.floor_value, 0, 100, "trust_reseed.floor_value"),
    ceilingValue: integer(reseedRow.ceiling_value, 0, 100, "trust_reseed.ceiling_value"),
  });
  if (trustReseed.floorValue > trustReseed.baseValue || trustReseed.baseValue > trustReseed.ceilingValue) syntax("trust_reseed bounds are invalid");
  const meters = root.meters.map((raw, meterIndex): MeterDefinition => {
    const label = `meter catalog.meters[${meterIndex}]`;
    const row = exactObject(raw, ["id", "scope", "min_value", "max_value", "initial_value", "bands", "inputs", "decay"], label);
    if (row.id !== REQUIRED_METER_IDS[meterIndex] || row.scope !== "company" || row.min_value !== METER_MIN_VALUE || row.max_value !== METER_MAX_VALUE || !Array.isArray(row.bands) || row.bands.length === 0 || !Array.isArray(row.inputs)) syntax(`${label} is invalid`);
    const bands: MeterBand[] = [];
    const seenBands = new Set<string>();
    let priorFloor = -1;
    for (let bandIndex = 0; bandIndex < row.bands.length; bandIndex += 1) {
      const bandRow = exactObject(row.bands[bandIndex], ["id", "floor_value"], `${label}.bands[${bandIndex}]`);
      const id = identifier(bandRow.id, `${label}.bands[${bandIndex}].id`);
      const floorValue = integer(bandRow.floor_value, 0, 100, `${label}.bands[${bandIndex}].floor_value`);
      if (seenBands.has(id) || floorValue <= priorFloor || bandIndex === 0 && floorValue !== 0) syntax(`${label}.bands are invalid`);
      seenBands.add(id); priorFloor = floorValue; bands.push(Object.freeze({ id, floorValue }));
    }
    const inputs: MeterInput[] = [];
    const seenInputs = new Set<string>();
    row.inputs.forEach((rawInput, inputIndex) => {
      const parsed = parseInput(rawInput, `${label}.inputs[${inputIndex}]`);
      if (seenInputs.has(parsed.uniqueness)) syntax(`${label}.inputs contain a duplicate source`);
      seenInputs.add(parsed.uniqueness); inputs.push(parsed.input);
    });
    let decay: MeterDecay | null = null;
    if (row.decay !== null) {
      const decayRow = exactObject(row.decay, ["toward_value", "rate_per_attended_hour"], `${label}.decay`);
      decay = Object.freeze({ towardValue: integer(decayRow.toward_value, 0, 100, `${label}.decay.toward_value`), ratePerHour: integer(decayRow.rate_per_attended_hour, 1, 100, `${label}.decay.rate_per_attended_hour`) });
    }
    return Object.freeze({ id: row.id as string, initialValue: integer(row.initial_value, 0, 100, `${label}.initial_value`), bands: Object.freeze(bands), inputs: Object.freeze(inputs), decay });
  });
  return Object.freeze({ schemaVersion: 1, trustReseed, meters: Object.freeze(meters), byId: new Map(meters.map((meter) => [meter.id, meter])) });
}

export function validateMeterResourceSeparation(catalog: MeterCatalog, resourceIds: readonly string[]): void {
  const meters = new Set(catalog.meters.map((meter) => meter.id));
  const collision = resourceIds.find((id) => meters.has(id));
  if (collision !== undefined) syntax(`meter/economy ID collision ${collision}`);
}
